### ISSUE TITLE
Unreliable behaviour of cv::fisheye::calibrate

### ISSUE THREAD

**Issue by ivashmak** [opencv/opencv#23962]
### System Information

OpenCV version: master.
Operating System: macOS Big Sur 11.7
Compiler: Apple clang version 12.0.0

### Detailed description

cv::fisheye::calibrate is very unreliable, it can either:
1. find intrinsics successfully - in most cases when CALIB_RECOMPUTE_EXTRINSIC flag is used
2. have error in initIntrinsics, ie (-215:Assertion failed) fabs(norm_u1) > 0 in function 'InitExtrinsics'
3. find intrinsics but retrojection is too high > 20 px - in most cases when other flags than CALIB_RECOMPUTE_EXTRINSIC are used.

Problems mainly found in Python. I have prepared two data files and code for reproducing the error:
In this test, for any flag, the error is:  (-215:Assertion failed) fabs(norm_u1) > 0 in function 'InitExtrinsics'. However, data contain good points detection (manually visualized and checked):
testFisheyeCalibrate('generated/points_InitExtrinsics_error.json')

In this test:
test_fisheye_calibrate_recompute_extrinsics_flags("points_default_flag_large_error.json");
Calibration works fine for CALIB_RECOMPUTE_EXTRINSIC, but fails (ie, has very large error) on default flag. Perhaps, there is some problem with default options, or CALIB_RECOMPUTE_EXTRINSIC should be added as default.

I have prepared Python/C++ code for reproducing the errors.

### Steps to reproduce

```python
import cv2 as cv
def testFisheyeCalibrate(file):
    read_file = cv.FileStorage(file, cv.FileStorage_READ)
    img_points = read_file.getNode('image_points').mat()
    obj_points = read_file.getNode('pattern_points').mat()
    flags = int(read_file.getNode('flags').real())
    print('flags', flags)
    image_size = (int(read_file.getNode('image_width').real()), int(read_file.getNode('image_height').real()))
    criteria = (cv.TERM_CRITERIA_EPS + cv.TERM_CRITERIA_MAX_ITER, 100, 2.22045e-16)
    repr_err_c, K, D, _, _ = cv.fisheye.calibrate([obj_points] * len(img_points), img_points, image_size, None, None, criteria=criteria, flags=flags)
    print('reprojection error', repr_err_c)

testFisheyeCalibrate('generated/points_InitExtrinsics_error.json')
testFisheyeCalibrate('generated/points_default_flag_large_error.json')
```

```cpp
#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/calib3d.hpp>

void test_fisheye_calibrate_recompute_extrinsics_flags(const std::string &fname) {
    cv::FileStorage file_storage (fname, cv::FileStorage::Mode::READ);
    CV_Assert(file_storage.isOpened());
    auto image_points_mat = file_storage["image_points"].mat();
    const int num_images = image_points_mat.size[0];
    cv::Mat pattern_points_mat = file_storage["pattern_points"].mat();
    const int num_pts = pattern_points_mat.size[0];
    const int flags = (int) file_storage["flags"].real();
    cv::Size img_size ((int) file_storage["image_width"].real(), (int) file_storage["image_height"].real());
    cv::Mat K, D, rvecs, tvecs;
    std::vector<cv::Mat> obj_points, img_points;
    obj_points.reserve(num_images);
    for (int i = 0; i < num_images; i++) {
        obj_points.emplace_back(pattern_points_mat);
    }
    img_points.reserve(num_images);
    for (int i = 0; i < num_images; i++) {
        img_points.emplace_back(image_points_mat.row(i).reshape(0, num_pts).reshape(2, num_pts));
    }
    cv::TermCriteria criteria = cv::TermCriteria(cv::TermCriteria::COUNT + cv::TermCriteria::EPS, 100, DBL_EPSILON);
    const double repr_err = cv::fisheye::calibrate(obj_points, img_points, img_size, K, D, rvecs, tvecs, flags, criteria);
    std::cout << "reprojection error " << repr_err << "\n" << K << "\n" << D << "\n";
    file_storage.release();
}

test_fisheye_calibrate_recompute_extrinsics_flags("points_InitExtrinsics_error.json");
test_fisheye_calibrate_recompute_extrinsics_flags("points_default_flag_large_error.json");
```
[data.zip](https://github.com/opencv/opencv/files/12009227/data.zip)

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by ivashmak** [comment]
@asmorkalov could you please have a look at this issue with fisheye calibration?

**Comment by ivashmak** [comment]
I was looking at the first error:  (-215:Assertion failed) fabs(norm_u1) > 0 in function 'InitExtrinsics'
And it seems that the problem is in ComputeHomography (https://github.com/opencv/opencv/blob/4.x/modules/calib3d/src/fisheye.cpp#L1296) that returns homography matrix filled with NaN values. I think it caused by numerical imprecision and division by 0, because in the internal loop some matrices have inf values.
I was wondering if we could use other methods to compute homography, eg, the least squares solution or RANSAC?

**Comment by ivashmak** [comment]
Regarding the second issue, I have found out that fisheye calibration often fails with the default flag, but works well for the CALIB_RECOMPUTE_EXTRINSIC flag.
The difference is that extrinsics are recomputed in the optimization loop (https://github.com/opencv/opencv/blob/4.x/modules/calib3d/src/fisheye.cpp#L836), with the default flag they remain constant. Therefore, the issue is that fisheye calibration relies too much on the initial guess of extrinsics parameters, and hence the optimization diverges.
Here, optimization steps for default flag:
iter 0 rmse test 11.2858 alpha 0.4
iter 1 rmse test 34.7766 alpha 0.64
iter 2 rmse test 66.2802 alpha 0.784
iter 3 rmse test 80.1968 alpha 0.8704
iter 4 rmse test 83.8994 alpha 0.92224

And here with CALIB_RECOMPUTE_EXTRINSIC flag:
iter 0 rmse test 11.2858
iter 1 rmse test 7.45222
iter 2 rmse test 3.45497
iter 3 rmse test 1.12238
iter 4 rmse test 0.488895

I suggest if we make CALIB_RECOMPUTE_EXTRINSIC option as default. Nevertheless, the extrinsics recomputation isn't cheap, so it's also possible to apply it every second or third iteration, which works well for my testing data.

More fundamentally, we could optimize not only over intrinsic parameters (as it is implemented now), but jointly over extrinsics + intrinsics, similarly how it is implemented for normal camera. I guess, this could be better for overall fisheye calibration accuracy. Otherwise, the current algorithm heavily relies on the extrinsics guess, and even recomputation of extrinsics may not give guarantee convergence.

**Comment by asmorkalov** [comment]
@ivashmak Thanks a lot for the investigation! Couple of proposals from my side:
- calibration pipeline is usually not time critical operation and executed only once. So, I propose to replace the flag with something like "FIX_EXTRINSICS" and always re-estimate extrinsics, if the flag is not set.
- It'll be great to find some public dataset with fisheye cameras and test the function robustness with real data. I can do it for couple of cameras I have, but it's not representative and I do not have ground truth to compare.

**Comment by lpanaf** [comment]
From my experiments, it seems that the current initialization of intrinsics can cause a large problem to the solver and was my cause to the error message: 

`(-215:Assertion failed) fabs(norm_u1) > 0 in function 'InitExtrinsics'`

I solved some instances by alternating the initialization of focal length. More specifically, I change it from `Vec2d(max(image_size.width, image_size.height) / CV_PI, max(image_size.width, image_size.height) / CV_PI)` to `Vec2d(max(image_size.width, image_size.height) / 2., max(image_size.width, image_size.height) / 2.)`.

To see the problem more concretely, assume that we have an observation point at the corner of the image, and it is an square image, then with the initialize parameter, the angle between the objective point and the camera principle axis would be $$\frac{\sqrt{2} \pi} {2}$$ which is larger than $\frac{\pi}{2}$. This means that the field of view would be larger than $180^\circ$ which is not supported by the opencv fisheye camera model (from formulation). In contrast, if initialize the parameter to be in the proposed form or to be 
``` 
double diag = std::sqrt(std::pow(image_size.width, 2) + std::pow(image_size.height, 2)) / 2;
Vec2d(diag / CV_PI, diag / CV_PI)
```
then, the field of view would be strictly smaller than $\pi$ and would be more stable.

**Comment by ivashmak** [comment]
I have also tried different initialization for intrinsic parameters in fisheye::calibrate. I agree with @lpanaf that current initialization could start with fairly large field of view, which results in numerical issues (ie, NaN/Inf values in homography matrix computation).

@lpanaf I tried your initialization, and for some 'calib3d' tests it results in this error:
"modules/calib3d/src/fisheye.cpp:1585: error: (-3:Internal error) CALIB_CHECK_COND - Ill-conditioned matrix for input array 2 in function 'CalibrateExtrinsics'"

I have experimented with another initialization (see below), where we start with the field of view that equals exactly 90 degrees.
```
            double focal_guess = max(image_size.width, image_size.height) / 2;
            finalParam.Init(Vec2d(focal_guess, focal_guess), ...
```
It works for my own instances, and for calib3d tests it results in slightly higher errors, for example:

### Failure
The max difference between matrices "theK" and "this->K" is 1.25608e-05 at (1, 2), which exceeds "1e-10", where "theK" at (1, 2) evaluates to (381.939), "this->K" at (1, 2) evaluates to (381.939), "1e-10" evaluates to 1e-10

### Failure
The max difference between matrices "theD" and "this->D" is 1.84127e-07 at (2, 0), which exceeds "1e-10", where "theD" at (2, 0) evaluates to (0.00605742), "this->D" at (2, 0) evaluates to (0.0060576), "1e-10" evaluates to 1e-10

### Failure
The max difference between matrices "err_std" and "cv::Vec2d(0.187475975266883, 0.185678953263995)" is 1.22326e-08 at (1, 0), which exceeds "1e-10", where "err_std" at (1, 0) evaluates to (0.185679), "cv::Vec2d(0.187475975266883, 0.185678953263995)" at (1, 0) evaluates to (0.185679), "1e-10" evaluates to 1e-10
 
So, the differences are lower than 1e-5, which is still good, but not reaching 1e-10 accuracy.

@asmorkalov if I may suggest keeping the current initialization, but during the first function's iteration we check whether there are numerical issues. If so, we repeat the whole process with a new initialization (either what @lpanaf suggested or mine).

**Comment by asmorkalov** [comment]
@ivashmak I guess that we can take down the test threshold. I do not know how the ground truth was generated and how it's close to reality. I suspect that it's just regression test and numbers were generated by OpenCV implementation some time ago. Several trials looks good to me too. Kalibr supports fisheye cameras. Matlab Computer Vision Toolbox supports fisheye cameras. Could you take a look what point do they use for the initial step?

**Comment by lpanaf** [comment]
@ivashmak I agree with your formulation, and this is exactly what I tried in my first experiments (`Vec2d(max(image_size.width, image_size.height) / 2., max(image_size.width, image_size.height) / 2.)`). The field of view of this is in fact larger than $90^\circ$ (2 rad, or $114.6^\circ$), and should be safe enough since the image ray of the points in the corner will at most be about $160^\circ$.

**Comment by ivashmak** [comment]
> @ivashmak I guess that we can take down the test threshold. I do not know how the ground truth was generated and how it's close to reality. I suspect that it's just regression test and numbers were generated by OpenCV implementation some time ago. Several trials looks good to me too. Kalibr supports fisheye cameras. Matlab Computer Vision Toolbox supports fisheye cameras. Could you take a look what point do they use for the initial step?

I also have feeling that the regression tests were set on OpenCV output.

- Matlab does not have open-source code, however, I found that they use implementation based on this paper: "A Toolbox for Easily Calibrating Omnidirectional Cameras" (Scaramuzza et al.), in the paper even said that the result is a Matlab Toolbox. The paper assumes omnidirectional camera model, and in the first step finds extrinsics parameters ($R_i, t_i$) for each frame $i$ to locate pattern grid. And in the second step, they solve set of linear equations to find intrinsics parameters directly. Afterwards, they polish results with linear refinement.

- Kalibr uses pinhole-equidistant camera model for corresponding in OpenCV fisheye model (see https://github.com/ethz-asl/kalibr/issues/256). They also say the same in the comments, here:
https://github.com/ethz-asl/kalibr/blob/master/aslam_cv/aslam_cameras/include/aslam/cameras/implementation/PinholeProjection.hpp#L713
This function does intrinsics initialization, the principal point is found similarly as in OpenCV (center of image), and focal length is found from vanishing points pairs.

**Comment by FunnyWii** [comment]
When I use my calibration code, I met the same issue.

Finally, I found the reason is the quality of my pics...I re-took a group of pics, and the calibration progress run successfully again.

**Comment by JohnnytheRocket** [comment]
I still have this issue in OpenCV 4.12.0 in C++ (running on Macbook Pro M3). With both a previously computed camera matrix and dist coefficients and not; as well as with calib_use_intrinsic_guess and not. It does work on the data set however if I increase the minimum detected markers to 7 for each image.

With OpenCV 4.4.0 I also encountered the problem with the same data set (with minimum detected markers >= 7), but after I had reverted a square procedure I was previously doing on the input image.

I guess that there are circumstances where some or a single detected marker can result in triggering the assertion error.
